cpp/src/server_handshake_context.cpp

```diff
@@ -0,0 +1,253 @@
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ * 
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ * 
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License
+ *
+ *****************************************************************************/
+
+#include <xtt-cpp/server_handshake_context.hpp>
+
+#include <stdexcept>
+
+using namespace xtt;
+
+server_handshake_context::server_handshake_context(unsigned char *in_buffer,
+                                                   uint16_t in_buffer_size,
+                                                   unsigned char *out_buffer,
+                                                   uint16_t out_buffer_size)
+{
+    xtt_return_code_type rc;
+    rc = xtt_initialize_server_handshake_context(&handshake_ctx_,
+                                                 in_buffer,
+                                                 in_buffer_size,
+                                                 out_buffer,
+                                                 out_buffer_size);
+    if (XTT_RETURN_SUCCESS != rc) {
+        throw std::runtime_error("Error initializing server handshake context");
+    }
+}
+
+server_handshake_context::server_handshake_context(server_handshake_context&& other)
+    : server_handshake_context(nullptr,
+                               0,
+                               nullptr,
+                               0)
+{
+    swap(*this, other);
+}
+
+server_handshake_context& server_handshake_context::operator=(server_handshake_context other)
+{
+    swap(*this, other);
+
+    return *this;
+}
+
+std::experimental::optional<version> server_handshake_context::get_version() const
+{
+    xtt_version current_version;
+    if (XTT_RETURN_SUCCESS != xtt_get_version(&current_version, &handshake_ctx_)) {
+        return {};
+    }
+
+    return static_cast<version>(current_version);
+}
+
+std::experimental::optional<suite_spec> server_handshake_context::get_suite_spec() const
+{
+    xtt_suite_spec current_suite_spec;
+    if (XTT_RETURN_SUCCESS != xtt_get_suite_spec(&current_suite_spec, &handshake_ctx_)) {
+        return {};
+    }
+
+    return static_cast<suite_spec>(current_suite_spec);
+}
+
+std::unique_ptr<pseudonym> server_handshake_context::get_clients_pseudonym() const
+{
+    switch (*get_suite_spec()) {
+        case suite_spec::X25519_LRSW_ED25519_CHACHA20POLY1305_SHA512:
+        case suite_spec::X25519_LRSW_ED25519_CHACHA20POLY1305_BLAKE2B:
+        case suite_spec::X25519_LRSW_ED25519_AES256GCM_SHA512:
+        case suite_spec::X25519_LRSW_ED25519_AES256GCM_BLAKE2B:
+            {
+                // Avoid double-copying
```

Comment:
Can't we avoid any copying here?

`xtt::pseudonym_lrsw` holds a `xtt_daa_pseudonym_lrsw` internally, so just allocate it first.

```
std::make_unique<pseudonym_lrsw> pseudo();
if (XTT_RETURN_SUCCESS == xtt_get_clients_pseudonum_lrsw(pseudo.get(), &handshake_ctx_))
   return std::move(pseudo);
else
   return {};
```